api/scpca_portal/s3.py

```diff
@@ -0,0 +1,148 @@
+
+    if recursive:
+        command_inputs.append("--recursive")
+
+    if "public" in str(bucket_path):
+        command_inputs.append("--no-sign-request")
+
+    try:
+        result = subprocess.run(command_inputs, capture_output=True, text=True, check=True)
+        output = result.stdout
+    except subprocess.CalledProcessError as error:
+        logger.error(
+            """
+            `{}`: Cause of error not returned, note: folder must exist and be non-empty
+            """.format(
+                error
+            )
+        )
+        return []
+
+    # The `aws s3 ls <bucket>` command returns a list of two types of entries:
+    # - Bucket Object Entries
+    # - Bucket Prefix Entries
+    # In order to create a standard API, where `entry.file_path` could be accessed
+    # irrespective of the entry type, we've created two named tuples which follow the return format
+    # of each of the bucket entry types.
+    BucketObjectEntry = namedtuple(
+        "BucketObjectEntry", ["date", "time", "size_in_bytes", "file_path"]
+    )
+    BucketPrefixEntry = namedtuple("BucketPrefixEntry", ["prefix_designation", "file_path"])
+    bucket_entries = []
+
+    for line in output.splitlines():
+        if line.strip().startswith("PRE"):
+            bucket_entries.append(BucketPrefixEntry._make(line.split()))
+        else:
+            bucket_entries.append(BucketObjectEntry._make(line.split()))
+
+    file_paths = [Path(entry.file_path) for entry in bucket_entries]
+
+    # recursive returns an absolute path (see docstring)
+    if recursive:
+        bucket_keys = Path(*bucket_path.parts[1:])
+        return [entry.relative_to(bucket_keys) for entry in file_paths]
+
+    return file_paths
+
+
+def download_input_files(file_paths: List[Path], *, bucket_name: str = common.INPUT_BUCKET_NAME):
+    """Download all passed data file paths which have not previously been downloaded.'"""
+    command_parts = ["aws", "s3", "sync", f"s3://{bucket_name}", common.INPUT_DATA_PATH]
+
+    download_queue = [fp for fp in file_paths if not fp.exists()]
+    # If download_queue is empty, exit early
+    if not download_queue:
+        return
+
+    command_parts.append("--exclude=*")
+    command_parts.extend([f"--include={file_path}" for file_path in download_queue])
+
+    if "public-test" in bucket_name:
+        command_parts.append("--no-sign-request")
+
+    subprocess.check_call(command_parts)
+
+
+def download_input_metadata(*, bucket_name: str = common.INPUT_BUCKET_NAME):
+    """Download all metadata files to the local file system."""
+    command_parts = ["aws", "s3", "sync", f"s3://{bucket_name}", common.INPUT_DATA_PATH]
+
+    command_parts.append("--exclude=*")
+    command_parts.append("--include=*_metadata.*")
+
+    if "public-test" in bucket_name:
+        command_parts.append("--no-sign-request")
+
+    subprocess.check_call(command_parts)
+
+
+def delete_output_file(key: str) -> bool:
```

Comment:
We should have a consistent API for all of the `upload`/`delete`/`download` `s3` module methods where they either return a `boolean` value or they don't. I understand the argument of handling (and logging) errors in the callee, and then allowing the caller to act based on the function's success or failure status, but we currently don't employ that type of error handling in our callers (at least where `s3::delete_output_file` is called in `Project::purge`).

How would you like to handle this?

Reply:
Anything that doesn't return a value should do this
- Use `try/catch` block inside of the s3 module functions

Don't need to adjust the caller for now

`list_s3_paths` - change the logging level from `error` to `warning`